File: src/main/java/com/leetcode/MaximumNumberOfVowelsInASubstringOfGivenLength.java
package com.leetcode;

/**
 * https://leetcode.com/problems/maximum-number-of-vowels-in-a-substring-of-given-length/submissions/1617254917/?envType=study-plan-v2&envId=leetcode-75
 */

public class MaximumNumberOfVowelsInASubstringOfGivenLength {

    public static void main(String[] args) {
        String s = "abciiidef";
        int k = 3;

        int result = maxVowels(s, k);
        System.out.println(result);
    }

    public static int maxVowels(String s, int k) {
        int cnt = 0;

        for(int i = 0 ; i < k ; i++){   // 초기 모음 갯수 계산
            char c = s.charAt(i);

            if(c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u') {
                cnt++;
            }
        }

        int maxCnt = cnt;   // 초기 모음 갯수를 최대 모음 갯수로 설정

        for(int i = k ; i < s.length() ; i++){  // 배열 확장하면서 최대 모음 갯수 갱신
            char beforeC = s.charAt(i-k);   // 이전 요소
            char currC = s.charAt(i);       // 현재 요소

            if(beforeC == 'a' || beforeC == 'e' || beforeC == 'i' || beforeC == 'o' || beforeC == 'u') {    // 이전 요소가 모음이면 갯수 1개 감소
                cnt--;
            }

            if(currC == 'a' || currC == 'e' || currC == 'i' || currC == 'o' || currC == 'u') {  // 현재 요소가 모음이면 갯수 1개 증가
                cnt++;
            }

            maxCnt = Math.max(maxCnt, cnt); // 최대 모음 갯수 갱신
        }

        return maxCnt;
    }
}

File: src/code/kotlin/MaximumNumberOfVowelsInASubstringOfGivenLength.kt
package code.kotlin

/**
 * https://leetcode.com/problems/maximum-number-of-vowels-in-a-substring-of-given-length/submissions/1617254917/?envType=study-plan-v2&envId=leetcode-75
 */

fun main(){
    val s = "abciiidef"
    val k = 3

    val result = maxVowels(s, k)
    println(result)
}

fun maxVowels(s: String, k: Int): Int {
    var cnt = 0

    for(i in 0 until k){    // 초기 모음 갯수
        val c = s.get(i)
        if(checkVowel(c)) cnt++
    }

    var max = cnt   // 초기 모음 갯수를 max로 설정

    for(i in k until s.length){     // 배열 확장하면서 모음 갯수 갱신
        // 이전 글자가 모음인 경우 cnt 1 감소
        val beforeC = s.get(i-k)
        if(checkVowel(beforeC)) cnt--

        // 현재 글자가 모음인 경우 cnt 1 증가
        var currC = s.get(i)
        if(checkVowel(currC)) cnt++

        max = maxOf(max, cnt)   // max와 cnt 비교하면서 갱신
    }

    return max
}

fun checkVowel(c: Char): Boolean {  // 모음 판단하는 메서드
    if(c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u'){
        return true
    } else {
        return false
    }
}